### PR TITLE
Somei70 addition

### PR DESCRIPTION
**Pull request by gzowski**
Somei70 - 70 key keyboard with OLED and RGB LED's

## Description

New keyboard to add to QMK. formatted json, kept everything in keyboard.json where possible. Only one layout being default but with VIA_ENABLE set. 
Not sure if this needed to be in it's own VIA keymap or whether default was fine.
Link to projects github currently private while build guide is being written up.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/somei70/keyboard.json
Comment (fauxpark) [inline keyboards/somei70/keyboard.json:1]: Please run `qmk format-json -i` on this file.

File: keyboards/somei70/keymaps/default/rules.mk [hunk 1]
```diff
@@ -0,0 +1,3 @@
+VIA_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/somei70/keymaps/default/rules.mk:1]:
```suggestion
```
VIA cannot be enabled on the default keymap.

File: keyboards/somei70/rules.mk
Comment (fauxpark) [inline keyboards/somei70/rules.mk:1]: This file can be deleted.

File: keyboards/somei70/keymaps/default/keymap.c
Comment (fauxpark) [inline keyboards/somei70/keymaps/default/keymap.c:1]: Please use consistent indentation.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/somei70/keyboard.json
Comment (fauxpark) [inline keyboards/somei70/keyboard.json:1]: Layout and RGB Matrix seems like it needs work.
image

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/somei70/keyboard.json
Comment (fauxpark) [inline keyboards/somei70/keyboard.json:1]: Layout is still wrong, please adjust the key sizes.
image
RGB Matrix doesn't make sense to me, are the LEDs only on the top row?

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/somei70/keyboard.json
Comment (gzowski) [reply keyboards/somei70/keyboard.json:1]: LED's only apply to top row, recalculated based on full keyboard width by 224, key positions should all now be correct. Checked appearence in qmk info -m -kb somei70 and all appears correct.

image

**Comment by gzowski** [comment]
Do I need to make any further changes?

File: keyboards/somei70/keymaps/default/rules.mk
Comment (fauxpark) [inline keyboards/somei70/keymaps/default/rules.mk:1]: This file can be deleted. OLED is enabled at keyboard level, and `ssd1306` is the default driver.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/somei70/keymaps/default/keymap.c
[the remark this answers is not in the document]
Comment (gzowski) [reply keyboards/somei70/keymaps/default/keymap.c:1]: removed tabs and formatted C for somei70.c

File: keyboards/somei70/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	KC_LCTL,KC_LGUI ,KC_LALT	,KC_SPC ,KC_MUTE,KC_ENT		,KC_RALT,MO(_LAYERB)	,KC_RCTL,KC_LEFT,KC_DOWN, KC_RGHT
+),
+	[_LAYERB] = LAYOUT_5x15(
+	KC_ESC ,KC_F1   ,KC_F2  ,KC_F3  ,KC_F4  ,KC_F5  ,KC_F6  ,KC_F7  ,KC_F8  ,KC_F9  ,KC_F10 ,KC_F11 ,KC_F12 ,KC_PSCR,
+	KC_TAB ,RGB_MOD ,RGB_VAD,RGB_VAI,RGB_SPD,RGB_SPI,RGB_HUD,RGB_HUI,RGB_SAD,RGB_SAI,KC_P   ,KC_LBRC,KC_RBRC,KC_BSLS,
+	KC_CAPS,KC_A    ,KC_S   ,KC_D   ,KC_F   ,KC_G   ,KC_H   ,KC_J   ,KC_K   ,KC_L   ,KC_SCLN,KC_QUOT,KC_GRV , KC_DEL, KC_PGUP,
+	KC_LSFT,KC_Z    ,KC_X   ,KC_C   ,KC_V   ,KC_B   ,KC_N   ,KC_M   ,KC_COMM,KC_DOT ,KC_SLSH,KC_RSFT,KC_INS , KC_UP , KC_PGDN,
+	KC_LCTL,KC_LGUI ,KC_LALT        ,KC_SPC ,RGB_TOG,KC_ENT         ,KC_RALT,_______	,KC_RCTL,KC_LEFT,KC_DOWN, KC_RGHT
+),
+	[_LAYERC] = LAYOUT_5x15(
```
[the remark this answers is not in the document]
Comment (gzowski) [reply keyboards/somei70/keymaps/default/keymap.c]: removed _LAYERC

File: keyboards/somei70/keyboard.json [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    "keyboard_name": "Somei70",
+    "maintainer": "gzowski",
+    "bootloader": "rp2040",
+    "diode_direction": "COL2ROW",
+    "features": {
+        "bootmagic": true,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true,
+        "oled": true,
```
[the remark this answers is not in the document]
Comment (gzowski) [reply keyboards/somei70/keyboard.json:12]: added somei70.c to include all OLED features.

File: keyboards/somei70/somei70.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            render_caps();
+        } else {
+            render_logo();
+        }
+    } else if (oled_mode == 1) {
+        render_stats();
+    }
+}
+// Rotate 180 degrees
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
```
Comment (waffle87) [inline keyboards/somei70/somei70.c:59]:
```suggestion
oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
```

File: keyboards/somei70/somei70.c [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    return OLED_ROTATION_180;
+}
+bool oled_task_user(void) {
```
Comment (waffle87) [inline keyboards/somei70/somei70.c:62]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false;
    }
```

File: keyboards/somei70/somei70.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    render_oled_display();
+    return false;
+}
+// Record keypress
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```
Comment (waffle87) [inline keyboards/somei70/somei70.c:67]:
```suggestion
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
```

File: keyboards/somei70/somei70.c [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+                if (esc_press_count == 3) {
+                    oled_clear();
+                    oled_mode       = (oled_mode == 0) ? 1 : 0;
+                    esc_press_count = 0;
+                }
+            }
+            break;
+    }
+    return true;
```
Comment (waffle87) [inline keyboards/somei70/somei70.c:85]:
```suggestion
    return process_record_user(keycode, record);
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Comment by gzowski** [comment]
Forgot I still had this open, Is there anything I'm missing out at this point to be added/modified?

File: keyboards/somei70/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  KC_CAPS, KC_A   , KC_S   , KC_D  , KC_F   , KC_G  , KC_H   , KC_J       , KC_K   , KC_L   , KC_SCLN, KC_QUOT , KC_GRV , KC_DEL , KC_PGUP,
+  KC_LSFT, KC_Z   , KC_X   , KC_C  , KC_V   , KC_B  , KC_N   , KC_M       , KC_COMM, KC_DOT , KC_SLSH, KC_RSFT , KC_INS , KC_UP  , KC_PGDN,
+  KC_LCTL, KC_LGUI, KC_LALT, KC_SPC, KC_MUTE, KC_ENT, KC_RALT, MO(_LAYERB), KC_RCTL, KC_LEFT, KC_DOWN, KC_RGHT
+),
+[_LAYERB] = LAYOUT_5x15(
+  KC_ESC , KC_F1  , KC_F2  , KC_F3  , KC_F4  , KC_F5  , KC_F6  , KC_F7  , KC_F8  , KC_F9  , KC_F10 , KC_F11 , KC_F12 , KC_PSCR,
+  KC_TAB , RGB_MOD, RGB_VAD, RGB_VAI, RGB_SPD, RGB_SPI, RGB_HUD, RGB_HUI, RGB_SAD, RGB_SAI, KC_P   , KC_LBRC, KC_RBRC, KC_BSLS,
+  KC_CAPS, KC_A   , KC_S   , KC_D   , KC_F   , KC_G   , KC_H   , KC_J   , KC_K   , KC_L   , KC_SCLN, KC_QUOT, KC_GRV , KC_DEL , KC_PGUP,
+  KC_LSFT, KC_Z   , KC_X   , KC_C   , KC_V   , KC_B   , KC_N   , KC_M   , KC_COMM, KC_DOT , KC_SLSH, KC_RSFT, KC_INS , KC_UP  , KC_PGDN,
+  KC_LCTL, KC_LGUI, KC_LALT, KC_SPC , RGB_TOG, KC_ENT , KC_RALT, _______, KC_RCTL, KC_LEFT, KC_DOWN, KC_RGHT
```
Comment (fauxpark) [inline keyboards/somei70/keymaps/default/keymap.c:18]:
```suggestion
  KC_TAB , RM_NEXT, RM_VALD, RM_VALU, RM_SPDD, RM_SPDU, RM_HUED, RM_HUEU, RM_SATD, RM_SATU, KC_P   , KC_LBRC, KC_RBRC, KC_BSLS,
  KC_CAPS, KC_A   , KC_S   , KC_D   , KC_F   , KC_G   , KC_H   , KC_J   , KC_K   , KC_L   , KC_SCLN, KC_QUOT, KC_GRV , KC_DEL , KC_PGUP,
  KC_LSFT, KC_Z   , KC_X   , KC_C   , KC_V   , KC_B   , KC_N   , KC_M   , KC_COMM, KC_DOT , KC_SLSH, KC_RSFT, KC_INS , KC_UP  , KC_PGDN,
  KC_LCTL, KC_LGUI, KC_LALT, KC_SPC , RM_TOGG, KC_ENT , KC_RALT, _______, KC_RCTL, KC_LEFT, KC_DOWN, KC_RGHT
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/somei70/somei70.c [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+#include QMK_KEYBOARD_H
+#ifdef OLED_ENABLE
+static uint8_t  oled_mode       = 0;
+static uint8_t  esc_press_count = 0;
+static uint16_t esc_timer       = 0;
+// Stats display
+void render_stats(void) {
+    char wpm_str[10];
+    snprintf(wpm_str, sizeof(wpm_str), "WPM: %03u", get_current_wpm());
+    oled_write_ln(wpm_str, false);
```
Comment (drashna) [inline keyboards/somei70/somei70.c:12]: I would recommend the `get_u8_str` function. It's much lighter on memory, and for stuff like this, recommended. 
```suggestion
    oled_write_P(PSTR("WPM: "), false);
    oled_write_ln(get_u8_str(get_current_wpm(), '0'), false);
```

File: keyboards/somei70/somei70.c [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    if (keymap_config.nkro) {
+        oled_write_ln("NKRO: On", false);
+    } else {
+        oled_write_ln("NKRO: Off", false);
+    }
+    uint8_t brightness = rgblight_get_val();
+    char    brightness_str[255];
+    snprintf(brightness_str, sizeof(brightness_str), "LED Brightness: %03u", brightness);
+    oled_write_ln(brightness_str, false);
```
Comment (drashna) [inline keyboards/somei70/somei70.c:26]:
```suggestion
    oled_write_P(PSTR("LED Brightness: "), false);
    oled_write_ln(get_u8_str(rgblight_get_val(), '0'), false);
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by waffle87 — APPROVED** [review]

**Pull request merged**